Adding new api version to devops RP

suyash691 opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet. Already live as preview (documentation is pending update hence this PR)
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

To initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

suyash691 committed:
Initial commit as per PR review

suyash691 committed:
Actual changes to api spec for new version

AutorestCI commented:
/azp run automation - sdk

suyash691 commented:
Already approved on private branch. Recreated PR as per recommendation from ARM team

suyash691 commented:
@chiragg4u Created new PR as per your recommendation here ->
https://github.com/Azure/azure-rest-api-specs/pull/11325

weidongxu-microsoft commented:
@suyash691 Let us know the approved private PR. That will save a lot of time.

suyash691 commented:
@weidongxu-microsoft https://github.com/Azure/azure-rest-api-specs-pr/pull/1648 This is the private PR url

weidongxu-microsoft commented:
@suyash691 
Thanks. Let me know when you plan to get this merged.

@pilor
This should be same as approved private PR Azure/azure-rest-api-specs-pr#1648 . Let me know if we can skip ARM review in this PR.

suyash691 commented:
@weidongxu-microsoft The changes in production to support this API are live so whenever @pilor and @bishal-pdMSFT confirm their approval, you can go ahead with the merge

suyash691 commented:
@weidongxu-microsoft Any updates on the merge?

weidongxu-microsoft commented:
/azp run

weidongxu-microsoft commented:
@suyash691 

There is still no response from ARM. Let me know the urgency or when you want it get merged. I might go ahead.

weidongxu-microsoft commented:
/azp run

weidongxu-microsoft commented:
@akning-ms Would you help merge this PR? API already online. ARM review passes.

The problem is that "SDK python" CI fails on build_package.

I've checked with Yuchao that this is not RP problem. But I cannot merge it since "SDK python" CI is required.

autorest pass
```
01:21:36.082 info 	13 changed files after SDK generate:
01:21:36.083 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/__init__.py
01:21:36.084 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/_configuration.py
01:21:36.085 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/_dev_ops_client.py
01:21:36.085 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/models/__init__.py
01:21:36.086 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/models/_dev_ops_client_enums.py
01:21:36.087 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/models/_models.py
01:21:36.087 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/models/_models_py3.py
01:21:36.088 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/models/_paged_models.py
01:21:36.089 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/operations/__init__.py
01:21:36.089 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/operations/_operations.py
01:21:36.090 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/operations/_pipeline_template_definitions_operations.py
01:21:36.090 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/operations/_pipelines_operations.py
01:21:36.093 info 		sdk/devops/azure-mgmt-devops/azure/mgmt/devops/version.py
01:21:36.093 git 	Commit all the changes
01:21:36.209 endsection 	Commit generate result
```

fail after
```
01:21:58.146 command 	python ./build_package.py --dest sdk/devops/azure-mgmt-devops sdk/devops/azure-mgmt-devops
01:21:58.146 info 	Config: {"exitCode":{"showInComment":true,"result":"error"},"stderr":{"showInComment":true},"logPrefix":"buildPackage"}
01:21:58.190 cmderr 	[buildPackage] Traceback (most recent call last):
01:21:58.191 cmderr 	[buildPackage]   File "./build_package.py", line 33, in <module>
01:21:58.191 cmderr 	[buildPackage]     create_package(args.name, args.dest)
01:21:58.191 cmderr 	[buildPackage]   File "./build_package.py", line 21, in create_package
01:21:58.191 cmderr 	[buildPackage]     absdirpath = os.path.abspath(absdirs[0])
01:21:58.191 cmderr 	[buildPackage] IndexError: list index out of range
```

bishal-pdMSFT commented on specification/devops/resource-manager/Microsoft.DevOps/preview/2020-07-13-preview/devops.json:
```diff
@@ -0,0 +1,920 @@
+      "properties": {
+        "sourceRepository": {
+          "$ref": "#/definitions/CodeRepository",
+          "description": "Repository containing the source code for the pipeline. Currently only 'azurePipeline' pipeline type supports this."
```
@suyash691 is there a way to enforce this using discriminator?

suyash691 replied:
I do not have a pipelineType or any property to switch between types from hence this will have to exist as is. It doesn't make sense from an OO model that we use sourceRepository as a descriminator

akning-ms commented:
Looks there is cross API version breaking change detected:
**[Staging] Cross Version BreakingChange (Base on preview version): 12 Warnings warning [Detail]**

suyash691 commented:
@akning-ms That is just a warning because the compact is not broken in the swagger spec but we have added new patterns to allow for inheritance/polymorphism which are flagging that. I do not think that should be breaking the python build

suyash691 committed:
Switching to multi-api build for python

suyash691 committed:
Fix python spec to correct namespaces in multiapi mode

suyash691 committed:
Revert changes to python generation

weidongxu-microsoft commented:
@akning-ms The python CI still fails at the same place, after autorest.

akning-ms commented:
> @akning-ms @weidongxu-microsoft That is just a warning because the compact is not broken in the swagger spec but we have added new patterns to allow for inheritance/polymorphism which are being flagged as warnings. I do not think that should be breaking the python build

Hi @suyash691, what I meant here is about API breaking change. Not python sdk build
compare with original: preview/2019-07-01-preview/devops.json <---> new: preview/2020-07-13-preview/devops.json
e.g repository rename to sourceRepository. 

For python sdk build fail. @PhoenixHe-msft  to help take a look

PhoenixHe-msft commented:
@msyyc Please help to investigate the python build failure

msyyc commented:
> @msyyc Yuchao Yan FTE Please help to investigate the python build failure

got it

weidongxu-microsoft commented on specification/devops/resource-manager/Microsoft.DevOps/preview/2020-07-13-preview/devops.json:
```diff
@@ -677,9 +727,9 @@
       "description": "Configuration used to bootstrap a Pipeline.",
       "properties": {
-        "repository": {
+        "sourceRepository": {
```
As @akning-ms mentioned, this change constitute as "remove" a property from last api-version. ("remove" then "add" a new one).

Though I am not sure how we treat this. The whole RP is in preview version. And this is cross api-version.

suyash691 replied:
@weidongxu-microsoft is there an action item on me for this?

suyash691 replied:
@weidongxu-microsoft @PhoenixHe-msft What is the action item on me for this? or do i need to wait for a set of reviewers to do the re-review?

suyash691 replied:
Also it's a removal and addition of property in a new API version which is still in preview. We are still supporting that property in the old API version, I don't think it should be considered a breaking change due to this

weidongxu-microsoft replied:
+ @JeffreyRichter for guidance on change of property name cross api-version on preview.

JeffreyRichter replied:
It is OK to add a property in a preview and then rename it during the preview. Our recommendation is that you version your previews. So, in api-version=2020-12-23-preview.1 you add a property. Then in api-version=2020-12-23-preview.2 you rename the property. If you follow this versioning scheme, then our tooling will not flag these changes as breaking. If you make these changes in the same api-version, then our tooling WILL flag these as breaking.

Also, when you release a new version, you can stop supporting the old preview version in 90 days. This gives your preview customers 90 days to move to the new preview without breaking them all of a sudden which can be frustrating to the customers that are trying to help you with giving you feedback on the preview(s).

JeffreyRichter replied:
Since you're in preview, I'll approve this PR but you have to notify your preview customers of the change, how it will impact them, and what timeframe they have to update their code by.

PhoenixHe-msft commented:
/azp run

suyash691 committed:
Add systemdata

suyash691 committed:
Fix the path for sys data reference

suyash691 committed:
Fix path once more

suyash691 committed:
remove additional properties after including systemdata

suyash691 committed:
Remove more addtional properties

suyash691 commented:
@weidongxu-microsoft @PhoenixHe-msft I added system data in the latest commit but still it is unable to detect system data in the response

weidongxu-microsoft commented:
@suyash691 
I've added Jeffrey for guidance on cross api-version change.

For system data, I think it need to be directly under `Pipeline` model, not its `properties` property. E.g. https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources

suyash691 committed:
Move system data to root

suyash691 commented:
> 
> 
> @suyash691
> I've added Jeffrey for guidance on cross api-version change.
> 
> For system data, I think it need to be directly under `Pipeline` model, not its `properties` property. E.g. https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources
Fixed the systemData related error
@weidongxu-microsoft  It seems that this time python CI didn't fail. Maybe the CI was has been patched to ignore cross version breaking changes (as it should) or does applying the "breakingchange" tag in the repo cause it to not run completely?

weidongxu-microsoft commented:
@suyash691

python CI build fail weeks ago is actually a bug on Swagger team. It is not your fault at all (that was why I pinged Arthur and Phoenix to merge/fix). I figure Swagger team fixed it last week.

The cross api-version validation is not very reliable, hence is reports as warning but not blocks the PR. We just being cautious on any removal of API/parameter/property. Since Jeffrey approved it you are OK to go.

Let me know when you need it merged.

suyash691 commented:
You can merge it today, we are ready with the changes on our RP
